x-pack/plugins/timelines/public/components/actions/timeline/cases/add_to_case_action.tsx

```diff
@@ -0,0 +1,317 @@
+    async (ev) => {
+      ev.preventDefault();
+      return navigateToApp('securitySolution', {
+        deepLinkId: SecurityPageName.case,
+        path: getCreateCaseUrl(urlSearch),
+      });
+    },
+    [navigateToApp, urlSearch]
+  );
+  const [isAllCaseModalOpen, openAllCaseModal] = useState(false);
+
+  const onCaseClicked = useCallback(
+    (theCase) => {
+      /**
+       * No cases listed on the table.
+       * The user pressed the add new case table's button.
+       * We gonna open the create case modal.
+       */
+      if (theCase == null) {
+        openCaseFlyoutOpen();
+      }
+      openAllCaseModal(false);
+    },
+    [openCaseFlyoutOpen]
+  );
+  const addNewCaseClick = useCallback(() => {
+    closePopover();
+    openCaseFlyoutOpen();
+  }, [openCaseFlyoutOpen, closePopover]);
+
+  const addExistingCaseClick = useCallback(() => {
+    closePopover();
+    openAllCaseModal(true);
+  }, [openAllCaseModal, closePopover]);
+
+  const items = useMemo(
+    () => [
+      <EuiContextMenuItem
+        key="add-new-case-menu-item"
+        onClick={addNewCaseClick}
+        aria-label={i18n.ACTION_ADD_NEW_CASE}
+        data-test-subj="add-new-case-item"
+        disabled={isDisabled}
+      >
+        <EuiText size="m">{i18n.ACTION_ADD_NEW_CASE}</EuiText>
+      </EuiContextMenuItem>,
+      <EuiContextMenuItem
+        key="add-existing-case-menu-item"
+        onClick={addExistingCaseClick}
+        aria-label={i18n.ACTION_ADD_EXISTING_CASE}
+        data-test-subj="add-existing-case-menu-item"
+        disabled={isDisabled}
+      >
+        <EuiText size="m">{i18n.ACTION_ADD_EXISTING_CASE}</EuiText>
+      </EuiContextMenuItem>,
+    ],
+    [addExistingCaseClick, addNewCaseClick, isDisabled]
+  );
+
+  const button = useMemo(
+    () => (
+      <EuiToolTip data-test-subj="attach-alert-to-case-tooltip" content={tooltipContext}>
+        <EuiButtonIcon
+          aria-label={ariaLabel}
+          data-test-subj="attach-alert-to-case-button"
+          size="s"
+          iconType="folderClosed"
+          onClick={openPopover}
+          isDisabled={isDisabled}
+        />
+      </EuiToolTip>
+    ),
+    [ariaLabel, isDisabled, openPopover, tooltipContext]
+  );
+
+  return (
+    <>
+      {userCanCrud && (
+        <ActionIconItem>
+          <EuiPopover
+            id="attachAlertToCasePanel"
+            button={button}
+            isOpen={isPopoverOpen}
+            closePopover={closePopover}
+            panelPaddingSize="none"
+            anchorPosition="downLeft"
+            repositionOnScroll
+          >
+            <EuiContextMenuPanel items={items} />
+          </EuiPopover>
+        </ActionIconItem>
+      )}
+      {isCreateCaseFlyoutOpen && (
+        <CreateCaseFlyout
+          afterCaseCreated={attachAlertToCase}
+          onCloseFlyout={closeCaseFlyoutOpen}
+          onSuccess={onCaseSuccess}
+        />
+      )}
+      {isAllCaseModalOpen &&
+        cases.getAllCasesSelectorModal({
```

Comment:
let's memoize this object and remove some of the new useMemo for `search` up there L142